src/lib/FreeC/Pass/KindInferencePass.hs

```diff
@@ -0,0 +1,211 @@
+--
+--   For simplification, all type variables have kind @*@. Therefore, all type
+--   constructors must be fully applied, so partial application of type
+--   constructors is not allowed. This pass also checks that type variables and
+--   function types do not occur on the left hand side of a type application.
+--   If a type expression is found that does not match this rules, a fatal
+--   error is thrown. For simplification, a type expression that satisfies all
+--   of these rules is called kind-correct.
+--
+--   = Example
+--
+--   The following module contains three invalid definitions
+--
+--   > module Failures where
+--   >
+--   > data MyList a = MyNil | MyCons a (MyList a)
+--   >
+--   > fail1 :: MyList (MyList a a) -> Integer
+--   > fail1 x = 42
+--   >
+--   > fail2 :: a a -> Integer
+--   > fail2 x = 42
+--   >
+--   > fail3 :: MyList -> Integer
+--   > fail3 x = 42
+--
+--   = Specification
+--
+--   == Preconditions
+--
+--   The environment must contain all type declarations that are used in the
+--   module
+--
+--   == Translations
+--
+--   The AST is not changed. It is checked if all type constructor applications
+--   have the right number of arguments. It is also checked whether there are
+--   type variables or function types on the left-hand side of type
+--   applications. In these cases, a fatal error is thrown. Otherwise, the
+--   pass finishes with the given module unchanged.
+--
+--   == Postcondition
+--
+--   All type constructors in type expressions are applied to the correct
+--   number of arguments and there are no type variables or function types on
+--   the left-hand side of type applications. 
+module FreeC.Pass.KindInferencePass
+  ( kindInferencePass
+  )
+where
+
+import           Control.Monad                  ( when )
+import           Data.Maybe                     ( fromMaybe )
+
+import           FreeC.Environment              ( lookupArity )
+import qualified FreeC.IR.Syntax               as IR
+import           FreeC.Monad.Converter
+import           FreeC.Monad.Reporter
+import           FreeC.Pass
+import           FreeC.Pretty                   ( showPretty )
+
+-- | A compiler pass that checks wheter there are type applications with type
+--   variables or function types on the left-hand side in the module and checks
+--   whether all type construcors in the module are fully applied.
+kindInferencePass :: Pass IR.Module
+kindInferencePass m@(IR.Module _ _ _ typeDecls typeSigs _ funcDecls) = do
+  checkTypeDecls typeDecls
+  checkTypeSigs typeSigs
+  checkFuncDecls funcDecls
+  return m
+
+-- |
+checkMaybeTypeSchema :: Maybe IR.TypeSchema -> Converter ()
+checkMaybeTypeSchema Nothing  = return ()
+checkMaybeTypeSchema (Just x) = checkTypeSchema x
+
+-- | Checks whether an optional type expression is kind correct in case the
+--   type expression is present
+checkMaybeType :: Maybe IR.Type -> Converter ()
+checkMaybeType Nothing  = return ()
+checkMaybeType (Just x) = checkType x
```

Comment:
`checkMaybeType` and `checkMaybeTypeSchema` are `mapM_ checkType` and `mapM_ checkTypeSchema`, respectively. You don't have to add a new function if it just applies `mapM_`.